### ISSUE TITLE
Method Proposal: pygame.geometry.Line.closest_point(point, do_clamp=True)

### ISSUE THREAD

**Issue by damusss** [pygame-community/pygame-ce#3019]
The `Line` object has not been ported yet, but for the time it will, I propose to add the method in the title.

The method returns the closest point on the line relative to another point. The math for it isn't super straight forward, and unless you studied that somehow you probably need to search it online, that's why I think it should be added.

The `do_clamp` parameter controls weather the endpoints are ignored. The "default" math returns a point that can exceed A and B, but this could be what the user wants, so it could be optionally clamped.

**Comment by bilhox** [comment]
I didn't understand what the feature does, it might be more elaboration and a schema.

**Comment by MrRedstone058** [comment]
> The `Line` object has not been ported yet, but for the time it will, I propose to add the method in the title.
> 
> The method returns the closest point on the line relative to another point. The math for it isn't super straight forward, and unless you studied that somehow you probably need to search it online, that's why I think it should be added.
> 
> The `do_clamp` parameter controls weather the endpoints are ignored. The "default" math returns a point that can exceed A and B, but this could be what the user wants, so it could be optionally clamped.

to understand what youre saying, is is basically finding the point in a line that is directly perpendicular to the point? (in the case of do_clamp = false, the closest point follows the gradient of the original line)?

**Comment by damusss** [comment]
@bilhox @MrRedstone058 I drawed what this method is supposed to do:
![image](https://github.com/user-attachments/assets/bf235846-1f26-4314-8b4f-af45c3df9979)
Now, with do_clamp = False, we have:
![image](https://github.com/user-attachments/assets/76b2b3ce-71b1-4f94-a683-f5ad72f019bb)
and with do_clamp = True, we have:
![image](https://github.com/user-attachments/assets/9ba9e85b-0064-4307-888d-53c450775f95)

**Comment by MrRedstone058** [comment]
yup, just as i thought. i do not know the best way to do the math for this, but i can try to work it out as a base

**Comment by bilhox** [comment]
After a discussion, we agreed to rename this function to ``Line.project``.
